src/System.Diagnostics.TraceSource/tests/CorrelationManagerTests.cs

```diff
@@ -43,11 +46,69 @@ public void CorrelationManager_EmptyStack()
         [Fact]
         public void CorrelationManager_ActivityId()
         {
-            Trace.CorrelationManager.ActivityId = Guid.Empty;
+        	Guid id = Guid.NewGuid();
+            Trace.CorrelationManager.ActivityId = id;
+
+            Assert.Equal(id, Trace.CorrelationManager.ActivityId);
+        }
+
+        [Fact]
+        public void CorrelationManager_ActivtyIdAcrossAwait()
+        {
+            Guid g = Guid.NewGuid();
+            Trace.CorrelationManager.ActivityId = g;
+            Assert.Equal(g, Trace.CorrelationManager.ActivityId);
+            Task.Yield().GetAwaiter().GetResult();
+            Assert.Equal(g, Trace.CorrelationManager.ActivityId);
+            Task.Delay(1).GetAwaiter().GetResult();
+            Assert.Equal(g, Trace.CorrelationManager.ActivityId);
+        }
+
+        [Fact]
+        public void CorrelationManager_CorrelationAcrossAwait()
+        {
+            Guid g = Guid.NewGuid();
+            Trace.CorrelationManager.ActivityId = g;
+
+
+            Assert.Equal(g, Trace.CorrelationManager.ActivityId);
+            Trace.CorrelationManager.StartLogicalOperation("one");
 
-            Assert.Equal(Guid.Empty, Trace.CorrelationManager.ActivityId);
+            // 1 operation in progress
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, "one");
+            Trace.CorrelationManager.StartLogicalOperation("two");
+            Task.Yield().GetAwaiter().GetResult();
+
+            // 2 operations in progress
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, "two", "one");
+            Trace.CorrelationManager.StopLogicalOperation();
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, "one");
+
+            Task.Delay(1).GetAwaiter().GetResult();
+            Trace.CorrelationManager.StopLogicalOperation();
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack);
         }
 
+
+        //[Fact]
+        //public void CorrelationManager_Concurrency()
+        //{
+        //    Trace.CorrelationManager.StartLogicalOperation();
+        //    bool run = true;
+        //    Task.WaitAll(Enumerable.Range(0, 10).Select(_ => Task.Run(() =>
+        //    {
+        //        try
+        //        {
+        //            while (Volatile.Read(ref run))
+        //            {
+        //                Trace.CorrelationManager.StartLogicalOperation(12345);
+        //                Trace.CorrelationManager.StopLogicalOperation();
+        //            }
+        //        }
+        //        finally { Volatile.Write(ref run, false); }
+        //    })).ToArray());
+        //}
```

Comment:
@justinvp I was using that to test some changes. I am in the process of rebasing and will remove this.